Docs for the dialogs (#15629)
ckeditor/ckeditor5: docs/framework/architecture/ui-library.md

To create a modal, use the optional {@link module:ui/dialog/dialog~DialogDefinition#isModal `isModal`} argument of the {@link module:ui/dialog/dialog~Dialog#show `Dialog#show()`} method:

```js
dialog.show( {
	isModal: true
} );
```

Always make sure there is a way for the user to close a modal - either via a "Close" button, <kbd>Esc</kbd> key or one of the action buttons.

#### Structure and behaviour

A {@link module:ui/dialog/dialogview~DialogView `DialogView`} can have three parts, each of which is optional: the header (which is used as a drag handler), the contents (the body of the dialog) and the actions area (a collection of buttons). The order of the parts can not be changed.

##### Header

A header may consist of any combination of three elements:

* an icon;
* a title;
* a "Close" button.

The "Close" button is added automatically as long as you provide an icon or a title. The code below shows how to create a header with an icon and a title, but without a "Close" button.

```js
dialog.show( {
	icon: pencilIcon,
	title: 'My first dialog',
	hasCloseButton: false
} );
```

##### Contents

> This part can be a single [view](#views) or a collection of views. They will be displayed directly inside a body of a dialog.

Reviewer: What I'm missing here is an example of how to inject any view into the dialog. The example in the first section is not helpful because it does not bring any content. I'd either extend the initial example or create another one with some content (e.g. simple paragraph view with some text) here.

The whole point of docs is to give developers a minimal snippet to achieve something at least somehow relateable to their goals. One thing I'm sure they'll want to display _something_ in their dialog.

Reviewer 2: +1. At least add some simple view with some text, like in manual tests.

Reviewer 2: I'd also add an information, here or somewhere, that the dialog size will depend on the view passed to it. Dialog itself does not have any settings related to size, so it is up to the integrator to create a view with a proper size.

Also, if the internal view size has changed it may be necessary to call `updatePositon()` -- I'd mention that. Unless we make this automatic. Which would be great.

			}
		},
		{
			label: 'Cancel',
			withText: true,
			onExecute: () => dialog.hide()
		}
	]
} );
```

#### Accessibility

Dialogs provide full keyboard accessiblity. While a dialog is open, press the <kbd>Ctrl</kbd>+<kbd>F6</kbd> combination to move the focus between the editor and the dialog. It can also be closed at any time by pressing the <kbd>Esc</kbd> key. To navigate through the dialog, use the <kbd>Tab</kbd> and <kbd>Shift</kbd>+<kbd>Tab</kbd> keystrokes. The contents of the dialog are also available for screen readers.

#### API

The dialog's lifecycle (creating and destroying) is managed by the {@link module:ui/dialog/dialog Dialog plugin}. It provides two public methods: {@link module:ui/dialog/dialog~Dialog#show `show()`} and {@link module:ui/dialog/dialog~Dialog#hide `hide()`}. Both of them fire the respective events ({@link module:ui/dialog/dialog~DialogShowEvent `show`} and {@link module:ui/dialog/dialog~DialogHideEvent `hide`}), so it's possible to hook after or before them.

##### `show()` method

This method accepts a number of configuration options that allow to shape the structure and behaviour of the dialog.

> 

Reviewer: DialogDefinition is worth mentioning here.

Reviewer 2: I'd say that an example is worth mentioning here.

In general, I am missing one, two or three nice examples. Maybe something simple, then the one with checkbox that I mentioned earlier.

);
```

### Dialogs and modals

Another UI component provided by the framework is a dialog. The dialog system in CKEditor 5 is brought by the {@link module:ui/dialog/dialog~Dialog `Dialog` plugin} that offers API for displaying [views](#views) in dialogs. In a sense, this plugin corresponds to another that manages views in balloons (popovers) across the UI ({@link module:ui/panel/balloon/contextualballoon~ContextualBalloon `ContextualBalloon` plugin}).

Dialog is a popup window that does not close when the user clicks outside of it. It allows for interacting with the editor and its content while being open (unless it is a modal, which blocks the interaction with the rest of the page until it is closed). When configured to display a [header](#header), a dialog is also {@link module:ui/bindings/draggableviewmixin~DraggableViewMixin draggable} using mouse or touch. Only one dialog can be open at a time - opening another one closes the previously visible one.

> A minimal plugin creating a toolbar button component showing an empty dialog can look like this:

Reviewer 2: ```suggestion
A minimal plugin that adds a toolbar button component which shows an empty dialog could look like this:
```

```js
import { Dialog, ButtonView } from '@ckeditor/ckeditor5-ui';

class MyPluginWithDialog extends Plugin {
	public static get requires() {
		return [ Dialog ] as const;
	}

	public init(): void {

> 		this.editor.ui.componentFactory.add( 'aiAssistant', locale => {

Reviewer 2: Change `aiAssistant` to something else.

To create a modal, use the optional {@link module:ui/dialog/dialog~DialogDefinition#isModal `isModal`} argument of the {@link module:ui/dialog/dialog~Dialog#show `Dialog#show()`} method:

```js
dialog.show( {
	isModal: true,

	// The rest of the dialog definition...
} );
```

There are different ways to close a modal: either by clicking the "Close" button in the corner (if the [header](#header) is visible), an <kbd>Esc</kbd> key press or one of the [action buttons](#action-buttons).

#### Structure and behaviour

A dialog can have three parts, each of which is optional: the [header](#header) (also used as a drag handler), the [content](#contents) (the body of the dialog) and the [action buttons](#action-buttons) area (a collection of buttons). The order of the parts can not be changed.

##### Header

A header may consist of any combination of three elements:

* the icon,
* the title,
* the "Close" button.

> The "Close" button is added automatically as long as you provide an icon or a title. The code below shows how to create display a dialog with a header that comes with an icon and a title, but without the "Close" button.

Reviewer 2: ```suggestion
By default, the "Close" button ("X") is added to the header as long as you provide an icon or a title. To hide it, set `hasCloseButton` flag to `false`:
```

Reviewer 2: Do we want to add:

*   some kind of info that if you hide the close button you should add a way to close the dialog (kind of obvious, I know),
*   somewhere information that the dialog can be closed using `Esc` key (and whether it still true if `hasCloseButton` is false),
    *   okay I see we have the info about `Esc` key in a11y section.
*   info how to prevent closing on `Esc`?
    *   or how to prevent closing in general (especiall in regards to modals)
    *   of info that this is a bad practice
*   info if model can be closed using `Esc`?
    *   this and above points are related

	title: 'My first dialog',
	// Do not display the "Close" button.
	hasCloseButton: false,

	// The rest of the dialog definition...
} );
```

##### Content

This part can be a single [view](#views) or a collection of views. They will be displayed directly inside a body of a dialog.

##### Action buttons

The last piece of the dialog is the actions area, where the buttons are displayed. Their behaviour can be fully customized using the `onCreate()` and `onExecute()` callbacks. Below you can find an example of the configuration for the four custom buttons:

* The "OK" button that closes the dialog and has a custom CSS class set,
* The "Set custom title" button that changes the dialog's title,
* The "Drag detector" button that changes its state once the dialog was dragged by the user.
* The "Cancel" button that closes the dialog.

```js
dialog.show( {
	actionButtons: [
		{
			label: 'OK',
			class: 'ck-button-action',
			withText: true,
			onExecute: () => dialog.hide()
		},
		{
			label: 'Set custom title',
			withText: true,
			onExecute: () => {
				dialog.view!.headerView!.label = 'New title';
			}
		},
		{
			label: 'Drag detector'
			withText: true,
			onCreate: buttonView => {

> 				buttonView.bind( 'isEnabled' ).to( dialog.view!, 'wasMoved', wasMoved => !wasMoved );

Reviewer 2: In other PR I asked whether `wasMoved` must be observable. I wonder if they are some real use cases where we need to react to the dialog `wasMoved` change (outside of the dialog itself).

Reviewer 2: If you want to change this example, you can add an example where button is disabled for some time after the dialog is opened.

Or even better example would be that the user has to click some kind of "I agree" checkbox and the button is active only then. But this will make this snippet less clear.

			}
		},
		{
			label: 'Cancel',
			withText: true,
			onExecute: () => dialog.hide()
		}
	]
} );
```

> 

Reviewer 2: I'd add a section talking about `onShow` and `onHide`. `onShow` could be with views actually, because you will usually set up something in the view after it is created, but not always, so maybe it can be in separate section.

#### Accessibility

Dialogs provide full keyboard accessibility. While a dialog is open, press the <kbd>Ctrl</kbd>+<kbd>F6</kbd> combination to move the focus between the editor and the dialog. It can also be closed at any time by pressing the <kbd>Esc</kbd> key. To navigate through the dialog, use the <kbd>Tab</kbd> and <kbd>Shift</kbd>+<kbd>Tab</kbd> keystrokes. The contents of the dialog are also available for screen readers.

#### API

The dialog's lifecycle (creating and destroying) is managed by the {@link module:ui/dialog/dialog Dialog plugin}. It provides two public methods: {@link module:ui/dialog/dialog~Dialog#show `show()`} and {@link module:ui/dialog/dialog~Dialog#hide `hide()`}. Both of them fire the respective events ({@link module:ui/dialog/dialog~DialogShowEvent `show`} and {@link module:ui/dialog/dialog~DialogHideEvent `hide`}), so it's possible to hook after or before them.

> 

Reviewer 2: It could be just in API docs but maybe it would be worth mentioning that the dialog is opened on `normal` priority and that `onShow` and `onHide` callbacks are called on `low` priority.

Also, now as I think of it, do we even need `onShow` and `onHide` if we have `show` and `hide` events? I guess they are simpler to use but I am getting redundancy smell. And then somebody reads guide and API docs and is confused when they should use this and when that. WDYT? cc @oleq

Reviewer: At some point, we decided to have both:
* callbacks for convenience: 90% of users will be satisfied by them and they won't need to dive into our events system, they are straightforward and correspond to the API in other dialog libraries.
* events for power users who want more

##### `show()` method

> This method accepts a number of configuration options that allow to shape the structure and behavior of the dialog.

Reviewer 2: Same way as you mentioned in `hide()` description, it's good to add here that this each time creates a new instance of a dialogview and shows it.

##### `show:[id]` event

When the {@link module:ui/dialog/dialog~Dialog#show `Dialog#show()`} function is called and the {@link module:ui/dialog/dialog~DialogDefinition#id `id`} parameter was provided, the fired event will be namespaced, allowing to customize some of the default dialog's properties. For example, you can change the default position of the "Find and replace" dialog from the editor corner to the bottom with the following code:

```js
dialog.on( 'show:findAndReplace', ( evt, data ) => {

> 	return Object.assign( data, { position: DialogViewPosition.EDITOR_CENTER } );

Reviewer 2: `return` is not needed here right? :thinking:

Reviewer 2: Description says "to the bottom" but used position moves it to the center ;).

</info-box>

#### Structure and behaviour

A dialog can have three parts, each of which is optional: the [header](#header) (also used as a drag handler), the [content](#content) (the body of the dialog) and the [action buttons](#action-buttons) area (a collection of buttons). The order of the parts can not be changed.

##### Header

A header may consist of any combination of three elements:

* the icon,
* the title,
* the "Close" button.

By default, the "Close" button ("X") is added to the header as long as you provide an icon or a title. To hide it, set `hasCloseButton` flag to `false`:

```js
import { icons } from 'ckeditor5/src/core.js';

// ...

editor.plugins.get( 'Dialog' ).show( {
	icon: icons.pencil,
	title: 'My first dialog',
	// Do not display the "Close" button.
	hasCloseButton: false,

	// The rest of the dialog definition...
} );
```

<info-box>
	If you decide to hide the "Close" button, remember to leave some other way(s) to close the dialog (like pressing the <kbd>Esc</kbd> key).
</info-box>

##### Content

> This part can be a single [view](#views) or a collection of views. They will be displayed directly inside a body of a dialog. You can find an example of how to insert a block of text into the dialog.

Reviewer 2: ```suggestion
This part can be any single [view](#views) or a collection of views. They will be displayed directly inside a body of a dialog. Below, you can find an example of how to insert a block of text into the dialog.
```

```js
const textView = new View( locale );

textView.setTemplate( {
	tag: 'div',
	attributes: {
		style: {
			padding: 'var(--ck-spacing-large)',
			whiteSpace: 'initial',
			width: '100%',
			maxWidth: '500px'
		},
		tabindex: -1
	},
	children: [
		'This is a sample content of the dialog.',
		'You can put here text, images, inputs, buttons, etc.'
	]
} );

editor.plugins.get( 'Dialog' ).show( {
	title: 'Info dialog with text',
	content: textView,

	// The rest of the dialog definition...
} );
```

##### Action buttons

The last piece of the dialog is the actions area, where the buttons are displayed. Their behaviour can be fully customized using the `onCreate()` and `onExecute()` callbacks. Below you can find an example of the configuration for the four custom buttons:

* The "OK" button that closes the dialog and has a custom CSS class set,
* The "Set custom title" button that changes the dialog's title,

> * The "Drag detector" button that changes its state once the dialog was dragged by the user.

Reviewer 2: Outdated description - the button was changed to something else.

* The "Cancel" button that closes the dialog.

```js
editor.plugins.get( 'Dialog' ).show( {
	// ...

	actionButtons: [
		{
			label: 'OK',
			class: 'ck-button-action',
			withText: true,
			onExecute: () => dialog.hide()
		},
		{
			label: 'Set custom title',
			withText: true,
			onExecute: () => {
				dialog.view!.headerView!.label = 'New title';
			}
		},
		{

> 			label: 'This button will be disabled in 3...'

Reviewer 2: I'd go the other way -- make it enabled in 3, 2, 1. It used to be a pattern in dialogs some time ago (so you won't click by mistake). And when enabled, change label to something like "Enabled after 3 seconds" or simply "Button".
